File: antsAtroposN4Command.R
library( ANTsR )

baseDirectory <- './'
dataDirectory <- paste0( baseDirectory, 'Images/' )

outputDirectory <- paste0( baseDirectory, 'OutputANTsR/' )
if( ! dir.exists( outputDirectory ) )
  {
  dir.create( outputDirectory )
  }
outputPrefix <- paste0( outputDirectory, 'antsr' )

numberOfOuterIterations <- 5

image <- antsImageRead( paste0( dataDirectory, 'KKI2009-01-MPRAGE_slice150.nii.gz' ), dimension = 2 )
mask <- antsImageRead( paste0( dataDirectory, 'KKI2009-01-MPRAGE_slice150_mask.nii.gz' ), dimension = 2 )
weightMask <- NULL

for( i in seq_len( numberOfOuterIterations ) )
  {
  cat( "***************   N4 <---> Atropos iteration ", i, "  ******************" )
  n4Results <- n4BiasFieldCorrection( img = image, mask = mask,
    weight_mask = weightMask, verbose = TRUE )
  image <- n4Results

  atroposResults <- atropos( a = image, x = mask, verbose = TRUE )

  # only use gm and wm probabilities for weight mask
  weightMask <- atroposResults$probabilityimages[[2]] *
                  ( 1 - atroposResults$probabilityimages[[1]] ) *
                  ( 1 - atroposResults$probabilityimages[[3]] ) +
                atroposResults$probabilityimages[[3]] *
                  ( 1 - atroposResults$probabilityimages[[1]] ) *
                  ( 1 - atroposResults$probabilityimages[[2]] )
  }

antsImageWrite( image, paste0( outputPrefix, "N4Corrected.nii.gz" ) )
antsImageWrite( atroposResults$segmentation, paste0( outputPrefix, "AtroposSegmentation.nii.gz" ) )

for( i in seq_len( length( atroposResults$probabilityimages ) ) )
  {
  antsImageWrite( atroposResults$probabilityimages[[i]],
    paste0( outputPrefix, "AtroposSegmentationProbability", i, ".nii.gz" ) )
  }

File: antsAtroposN4Command.py
import ants
import os

baseDirectory = './'
dataDirectory = baseDirectory + 'Images/'

outputDirectory = baseDirectory + 'OutputANTsPy/'
if not os.path.isdir( outputDirectory ):
  os.mkdir( outputDirectory )

outputPrefix = outputDirectory + 'antspy'

numberOfOuterIterations = 5

image = ants.image_read( dataDirectory + 'KKI2009-01-MPRAGE_slice150.nii.gz', dimension = 2 )
mask = ants.image_read( dataDirectory + 'KKI2009-01-MPRAGE_slice150_mask.nii.gz', dimension = 2 )
weightMask = None

for i in range( numberOfOuterIterations ):
  print( "***************   N4 <---> Atropos iteration ", i, "  ******************\n" )
  n4Results = ants.n4_bias_field_correction( image, mask = mask,
    weight_mask = weightMask, verbose = True )
  image = n4Results
  atroposResults = ants.atropos( a = image, x = mask )
  onesImage = ants.make_image( image.shape, voxval = 0,
    spacing = ants.get_spacing( image ),
    origin = ants.get_origin( image ),
    direction = ants.get_direction( image ) )
  weightMask = atroposResults['probabilityimages'][1] *\
    ( onesImage - atroposResults['probabilityimages'][0] ) *\
    ( onesImage - atroposResults['probabilityimages'][2] ) +\
    atroposResults['probabilityimages'][2] *\
    ( onesImage - atroposResults['probabilityimages'][1] ) *\
    ( onesImage - atroposResults['probabilityimages'][0] )

ants.image_write( image, outputPrefix + "N4Corrected.nii.gz" )
ants.image_write( atroposResults['segmentation'], outputPrefix + "AtroposSegmentation.nii.gz" )

for i in range( len( atroposResults['probabilityimages'] ) ):
  ants.image_write( atroposResults['probabilityimages'][i],
    outputPrefix + "AtroposSegmentationProbability" + str( i ) + ".nii.gz" )

